sql/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectExecutionManager.scala

```diff
@@ -108,43 +110,50 @@ private[connect] class SparkConnectExecutionManager() extends Logging {
    * execution if still running, free all resources.
    */
   private[connect] def removeExecuteHolder(key: ExecuteKey, abandoned: Boolean = false): Unit = {
-    var executeHolder: Option[ExecuteHolder] = None
+    val executeHolder = executions.get(key)
+
+    if (executeHolder == null) {
+      return
+    }
+
+    // Put into abandonedTombstones before removing it from executions, so that the client ends up
+    // getting an INVALID_HANDLE.OPERATION_ABANDONED error on a retry.
+    if (abandoned) {
+      abandonedTombstones.put(key, executeHolder.getExecuteInfo)
+    }
+
+    // Remove the execution from the map *after* putting it in abandonedTombstones.
+    executions.remove(key)
+    executeHolder.sessionHolder.removeExecuteHolder(executeHolder.operationId)
+
     executionsLock.synchronized {
-      executeHolder = executions.remove(key)
-      executeHolder.foreach { e =>
-        // Put into abandonedTombstones under lock, so that if it's accessed it will end up
-        // with INVALID_HANDLE.OPERATION_ABANDONED error.
-        if (abandoned) {
-          abandonedTombstones.put(key, e.getExecuteInfo)
-        }
-        e.sessionHolder.removeExecuteHolder(e.operationId)
-      }
       if (executions.isEmpty) {
         lastExecutionTimeMs = Some(System.currentTimeMillis())
       }
-      logInfo(log"ExecuteHolder ${MDC(LogKeys.EXECUTE_KEY, key)} is removed.")
     }
-    // close the execution outside the lock
-    executeHolder.foreach { e =>
-      e.close()
-      if (abandoned) {
-        // Update in abandonedTombstones: above it wasn't yet updated with closedTime etc.
-        abandonedTombstones.put(key, e.getExecuteInfo)
-      }
+
+    logInfo(log"ExecuteHolder ${MDC(LogKeys.EXECUTE_KEY, key)} is removed.")
+
+    executeHolder.close()
+    if (abandoned) {
+      // Update in abandonedTombstones: above it wasn't yet updated with closedTime etc.
+      abandonedTombstones.put(key, executeHolder.getExecuteInfo)
     }
   }
 
   private[connect] def getExecuteHolder(key: ExecuteKey): Option[ExecuteHolder] = {
-    executionsLock.synchronized {
-      executions.get(key)
-    }
+    Option(executions.get(key))
   }
 
   private[connect] def removeAllExecutionsForSession(key: SessionKey): Unit = {
-    val sessionExecutionHolders = executionsLock.synchronized {
-      executions.filter(_._2.sessionHolder.key == key)
-    }
-    sessionExecutionHolders.foreach { case (_, executeHolder) =>
+    var sessionExecutionHolders = mutable.ArrayBuffer[ExecuteHolder]()
+    executions.forEach((_, executeHolder) => {
+      if (executeHolder.sessionHolder.key == key) {
+        sessionExecutionHolders += executeHolder
+      }
+    })
+
+    sessionExecutionHolders.foreach { executeHolder =>
```

Comment:
There's a comment in SessionHolder:
```
    // Clean up all executions.
    // After closedTimeMs is defined, SessionHolder.addExecuteHolder() will not allow new executions
    // to be added for this session anymore. Because both SessionHolder.addExecuteHolder() and
    // SparkConnectExecutionManager.removeAllExecutionsForSession() are executed under
    // executionsLock, this guarantees that removeAllExecutionsForSession triggered here will
    // remove all executions and no new executions will be added in the meanwhile.
    SparkConnectService.executionManager.removeAllExecutionsForSession(this.key)
```
which needs to be adjusted (and thought through for correctness now...).

It's no longer under executionsLock, but removeAllExecutionsForSession is called after closedTimeMs is already defined...
If there is a `sessionHolder.addExecuteHolder(executeHolder)` from https://github.com/apache/spark/pull/48034/files#diff-a1ce7b58be6b0e0cd79bea6a5ee2b399d48d18e90e0ae50e1d3160d74f639ad7R96 that's in progress that started when closedTimeMs was not yet defined, then that executeHolder has already been put into the global list in https://github.com/apache/spark/pull/48034/files#diff-a1ce7b58be6b0e0cd79bea6a5ee2b399d48d18e90e0ae50e1d3160d74f639ad7R85, then it will be removed here by removeAllExecutionsForSession
So I think it's fine... Think about it.

Reply:
Ah, that's a valid point, thanks! So, the following execution sequence will be really problematic.

Thread 1: check closedTimeMs -> None.
Thread 2: set closedTimeMs.
Thread 2: removeAllExecutionsForSession.
Thread 1: add the executeHolder to the map.

Furthermore, closedTimeMs is a plain Option, which is *semantically* not quite thread-safe (it will be 16-byte long so if not perfectly aligned in memory, it is true).

I'll try to fix those issues in this PR too!

Reply:
Ignore the thread-safety of closedTimeMs -> it is marked volatile so should be safe.

Reply:
```
1. Thread 1: check closedTimeMs -> None.
2. Thread 2: set closedTimeMs.
3. Thread 2: removeAllExecutionsForSession.
4. Thread 1: add the executeHolder to the map.
```

I think this is fine, because in 4. it will only add the executeHolder to the map in SessionHolder, and we’re throwing that SessionHolder away at that point.
I think it is not possible to add it to the global map in ExecutionManager at that point, because in 1. it was already added there.

Reply:
You're right, this will be fine.